File: python/tree/leetocde/valid_bst.py
# 验证二叉搜索树

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    def __init__(self):
        self.prev = None

    # 中序遍历，然后看数据是否完全有序
    def isValidBST_1(self, root: TreeNode) -> bool:
        def inOrder(node: TreeNode):
            if not node:
                return
            inOrder(node.left)
            res.append(node.val)
            inOrder(node.right)

        res = []
        inOrder(root)
        return sorted(set(res)) == res

    # 中序遍历，无需保存全部的数据，前后数据比较即可
    def isValidBST_2(self, root: TreeNode) -> bool:
        if not root:
            return True
        if not self.isValidBST_2(root.left):
            return False
        if self.prev and self.prev.val >= root.val:
            return False
        self.prev = root
        return self.isValidBST_2(root.right)

    # 递归的方式
    def isValidBST_3(self, root: TreeNode) -> bool:
        def helper(node: TreeNode, mi, ma):
            if not node:
                return True
            if node.val >= ma or node.val <= mi:
                return False
            return helper(node.left, mi, node.val) and helper(node.right, node.val, ma)

        return helper(root, float('-inf'), float('inf'))

File: go/tree/leetcode/valid_bst.go
package leetcode

import (
	"math"
	"reflect"
	"sort"
)

//中序遍历，并排序比较
func isValidBST1(root *TreeNode) bool {
	var (
		res  []int
		temp []int
	)
	set := map[int]bool{}
	helper1(root, set, &res)

	for k := range set {
		temp = append(temp, k)
	}
	sort.Ints(temp)
	return reflect.DeepEqual(temp, res)
}

func helper1(node *TreeNode, set map[int]bool, res *[]int) {
	if node != nil {
		helper1(node.Left, set, res)
		set[node.Val] = true
		*res = append(*res, node.Val)
		helper1(node.Right, set, res)
	}
}

var prev *TreeNode

//中序遍历优化，比较前后节点
func isValidBST2(root *TreeNode) bool {
	if root == nil {
		return true
	}

	if !isValidBST2(root.Left) {
		return false
	}
	if prev != nil && prev.Val >= root.Val {
		return false
	}
	prev = root
	return isValidBST2(root.Right)
}

func isValidBST3(root *TreeNode) bool {
	return helper3(root, math.MinInt64, math.MaxInt64)
}

//递归比较
func helper3(node *TreeNode, min, max int) bool {
	if node == nil {
		return true
	}
	if node.Val <= min || node.Val >= max {
		return false
	}
	return helper3(node.Left, min, node.Val) && helper3(node.Right, node.Val, max)
}
